### ISSUE TITLE
Extract gql extraction into another package

### ISSUE THREAD

**Issue by lewisf** [apollographql/persistgraphql#19]
The GQL extraction provided in persistgraphql is pretty great. I'm hoping to use it to add support for javascript source files within apollo-codegen or similar projects and having the gql extraction stuff available as a separate package would make that much easier. It that something that sounds reasonable?

**Comment by Poincare** [comment]
Yeah, definitely! In fact, that would help solve #17 as well since people could just build stuff on top of the `ExtractGQL` class or similar. Would be happy to accept PRs on this and also get involved in the implementation.

**Comment by lewisf** [comment]
awesome, this is something that im happy to take a look at in the next week or so.

where should this go? should an apollographql/extract-gql repo be created for this?

**Comment by stubailo** [comment]
@lewisf I think the best option is for you to create a prototype on a personal repo, then we can move it into the organization afterwards?

We should also see if this can replace https://github.com/apollographql/graphql-document-collector, which is pretty old and I'm not sure many people use it. The intention was to do something similar.

**Comment by lewisf** [comment]
Okay, sounds reasonable, I'll give it a shot

**Comment by lewisf** [comment]
Quick update after speaking with @Poincare. It looks like the ExtractGQL class is exactly what should get extracted.

After extracting it we'd be able to use it to solve #17 as mentioned above, and I'd be able to use it to address https://github.com/apollographql/apollo-codegen/issues/25.

I'm going to work on pulling it out into another package and seeing if anything needs to change (public/private methods) etc. We can leave this issue open until that package exists and is integrated back into persistgraphql.

**Comment by lewisf** [comment]
Okay, I've done some initial work in extracting ExtractGQL here:
https://github.com/lewisf/extract-gql

I've even tried using it in another project locally and it worked well (yay). I haven't done an audit of the public interface but it's good to know that the extraction was pretty straightforward due to how ExtractGQL was written here.

Some key improvements I want to make to extract-gql before publishing it as a package are going to be documented there. Namely:

- https://github.com/lewisf/extract-gql/issues/1
- https://github.com/lewisf/extract-gql/issues/2
- https://github.com/lewisf/extract-gql/issues/3

Please let me know if there are any other issues that should be addressed before extract-gql should go live.

I've also opened up other issues on it based on me trying to use it with an existing project we have here at Coursera, but I don't think those are blocking the re-integration of [extract-gql](https://github.com/lewisf/extract-gql) back into persistgraphql.

**Comment by Poincare** [comment]
@lewisf Great! Most of those issues and general direction look good to me. 

However, I did think that the package you would be building wouldn't actually include any of the CLI portions that this tool provides. Instead, it should just focus on providing a simple library that allows other projects (e.g. `apollo-codegen`) to build on top of the query extraction. Then, `persistgraphql` can simply use that library internally rather than having the `ExtractGQL` class as it is currently implemented and `persistgraphql` would provide the CLI functionality on top. Does that sound reasonable?

**Comment by lewisf** [comment]
@Poincare yes that definitely does sound reasonable. I'll go ahead and make that modification to extract-gql's responsibilities

**Comment by Poincare** [comment]
Hey @lewisf any progress on this? Any way I can help?

**Comment by lewisf** [comment]
@Poincare sorry! i had to shift focuses to prioritize something else -- happy to chat about this if you wanted to take a stab at this
